url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/decomposition/eig/watched/WatchedDoubleStepQREigen.java

```diff
@@ -0,0 +1,602 @@
+        double a21 = A.get(x1+1,x1);
+        double a12 = A.get(x1,x1+1);
+        double a22 = A.get(x1+1,x1+1);
+        double a32 = A.get(x1+2,x1+1);
+
+        double p_plus_t = 2.0*real;
+        double p_times_t = real*real + img*img;
+
+        double b11,b21,b31;
+        if( useStandardEq ) {
+            b11 = (a11*a11 - p_plus_t*a11+p_times_t)/a21 + a12;
+            b21 = a11+a22-p_plus_t;
+            b31 = a32;
+        } else {
+            // this is different from the version in the book and seems in my testing to be more resilient to
+            // over flow issues
+            b11 = (a11*a11 - p_plus_t*a11+p_times_t) + a12*a21;
+            b21 = (a11+a22-p_plus_t)*a21;
+            b31 = a32*a21;
+        }
+
+        performImplicitDoubleStep(x1, x2, b11, b21, b31);
+    }
+
+    private void performImplicitDoubleStep(int x1, int x2,
+                                           double b11 , double b21 , double b31 ) {
+        if( !bulgeDoubleStepQn(x1,b11,b21,b31,0,false) )
+            return;
+
+        // get rid of the bump
+        if( Q != null ) {
+            QrHelperFunctions.rank1UpdateMultR(Q,u.data,gamma,0,x1,x1+3,_temp.data);
+            if( checkOrthogonal && !MatrixFeatures.isOrthogonal(Q,1e-8) ) {
+                u.print();
+
+                Q.print();
+                throw new RuntimeException("Bad");
+            }
+        }
+
+        if( printHumps ) {
+            System.out.println("Applied first Q matrix, it should be humped now. A = ");
+            A.print("%12.3e");
+            System.out.println("Pushing the hump off the matrix.");
+        }
+
+        // perform double steps
+        for( int i = x1; i < x2-2; i++ ) {
+            if( bulgeDoubleStepQn(i) && Q != null ) {
+                QrHelperFunctions.rank1UpdateMultR(Q,u.data,gamma,0,i+1,i+4,_temp.data);
+                if( checkOrthogonal && !MatrixFeatures.isOrthogonal(Q,1e-8) )
+                    throw new RuntimeException("Bad");
+            }
+
+            if( printHumps ) {
+                System.out.println("i = "+i+" A = ");
+                A.print("%12.3e");
+            }
+        }
+        if( printHumps )
+            System.out.println("removing last bump");
+        // the last one has to be a single step
+        if( x2-2 >= 0 && bulgeSingleStepQn(x2-2) && Q != null ) {
+            QrHelperFunctions.rank1UpdateMultR(Q,u.data,gamma,0,x2-1,x2+1,_temp.data);
+            if( checkOrthogonal && !MatrixFeatures.isOrthogonal(Q,1e-8) )
+                throw new RuntimeException("Bad");
+
+        }
+        if( printHumps ) {
+            System.out.println(" A = ");
+            A.print("%12.3e");
+        }
+//        A.print("%12.3e");
+
+        if( checkHessenberg && !MatrixFeatures.isUpperTriangle(A,1,1e-12)) {
+            A.print("%12.3e");
+            throw new RuntimeException("Bad matrix");
+        }
+    }
+
+
+    public void performImplicitSingleStep(int x1, int x2 , double eigenvalue ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 342-364 into a separate method.